bindings/rust/bench/src/rustls.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    harness::{
+        read_to_bytes, CipherSuite, ConnectedBuffer, CryptoConfig, ECGroup, Mode, TlsBenchHarness,
+    },
+    CA_CERT_PATH, SERVER_CERT_CHAIN_PATH, SERVER_KEY_PATH,
+};
+use rustls::{
+    cipher_suite::{TLS13_AES_128_GCM_SHA256, TLS13_AES_256_GCM_SHA384},
+    kx_group::SECP256R1,
+    version::TLS13,
+    Certificate, ClientConfig, ClientConnection, PrivateKey,
+    ProtocolVersion::TLSv1_3,
+    RootCertStore, ServerConfig, ServerConnection, ServerName,
+};
+use rustls_pemfile::{certs, pkcs8_private_keys};
+use std::{error::Error, io::BufReader, sync::Arc};
+
+pub struct RustlsHarness {
+    client_buf: ConnectedBuffer,
+    server_buf: ConnectedBuffer,
+    client_conn: ClientConnection,
+    server_conn: ServerConnection,
+}
+
+impl RustlsHarness {
+    fn get_root_cert_store() -> Result<RootCertStore, Box<dyn Error>> {
+        let root_cert =
+            Certificate(certs(&mut BufReader::new(&*read_to_bytes(CA_CERT_PATH)))?.remove(0));
+        let mut root_certs = RootCertStore::empty();
+        root_certs.add(&root_cert)?;
+        Ok(root_certs)
+    }
+
+    fn get_cert_chain() -> Result<Vec<Certificate>, Box<dyn Error>> {
+        let chain = certs(&mut BufReader::new(&*read_to_bytes(SERVER_CERT_CHAIN_PATH)))?;
+        Ok(chain
+            .iter()
+            .map(|bytes| Certificate(bytes.to_vec()))
+            .collect())
+    }
+
+    fn get_server_key() -> Result<PrivateKey, Box<dyn Error>> {
+        Ok(PrivateKey(
+            pkcs8_private_keys(&mut BufReader::new(&*read_to_bytes(SERVER_KEY_PATH)))?.remove(0),
+        ))
+    }
+
+    /// Read all incoming data, process it, and write it out
+    /// Works for handshaking and data transfer
+    fn process_conn(&mut self, mode: Mode) -> Result<(), std::io::Error> {
+        match match mode {
+            Mode::Client => self.client_conn.complete_io(&mut self.client_buf),
+            Mode::Server => self.server_conn.complete_io(&mut self.server_buf),
+        } {
+            Ok(_) => Ok(()),
+            Err(err) => {
+                if err.kind() == std::io::ErrorKind::WouldBlock {
```

Comment:
Do we ever hit this? From the `complete_io` documentation is looks like it will only try to read if the TLS protocol wants to read, in which case we might be able to get away without it?

https://docs.rs/rustls/latest/rustls/struct.ConnectionCommon.html#method.complete_io

Reply:
We do hit this. Like with above how we have to return a blocking error on read of length 0 (no data available), Rustls  blocks because IO blocks.